src/tool/subcommands/car_cmd.rs

```diff
@@ -50,7 +69,51 @@ impl CarCommands {
                 crate::db::car::forest::Encoder::write(&mut writer, all_roots, frames).await?;
                 writer.flush().await?;
             }
+            Self::Validate {
+                car_file,
+                ignore_block_validity: check_block_validity,
+                ignore_forest_index: check_forest_index,
+            } => validate(car_file, check_block_validity, check_forest_index).await?,
         }
         Ok(())
     }
 }
+
+// At present, three invariants are checked:
+// - The CAR file is syntactically valid and all blocks can be streamed.
+// - Each block CID is checked against the hash of the block.
+// - Each block CID is looked-up in the on-disk index.
+//
+// Invariants related to Filecoin are not checked. For those, see `forest-tool
+// snapshot validate`.
+//
+// We do not check for duplicate blocks. Whether duplicate blocks are allowed or
```

Comment:
Should we have some tests for this feature? Or in compiler and infallible code we trust?

Reply:
Done. All the error cases are now tested.